wandb/sdk/internal/internal_api.py

```diff
@@ -1552,6 +1549,32 @@ def download_write_file(self, metadata, out_dir=None):
 
         return path, response
 
+    def upload_file_azure(self, url, file, extra_headers):
+        from azure.core.exceptions import HttpResponseError
+
+        client = self._azure_blob_module.BlobClient.from_blob_url(url)
+        try:
+            if extra_headers.get("Content-MD5") is not None:
+                md5 = base64.b64decode(extra_headers["Content-MD5"])
+            else:
+                md5 = None
+            content_settings = self._azure_blob_module.ContentSettings(
+                content_md5=md5, content_type=extra_headers.get("Content-Type"),
+            )
+            client.upload_blob(
+                file,
+                max_concurrency=4,
```

Comment:
Why this limit? MS [promises](https://docs.microsoft.com/en-us/azure/developer/python/sdk/storage/azure-storage-blob/azure.storage.blob.containerclient?view=storage-py-v12#upload-blob-name--data--blob-type--blobtype-blockblob---blockblob----length-none--metadata-none----kwargs-) optimal performance with automated chunking given accurate length.

Reply:
The default for the library is 1.  It only uploads chunks of a blob concurrently if the blob is larger than 64MB.  Since we're already uploading files concurrently I figured this was a reasonable default.